File: project/jimmer-sql/src/test/java/org/babyfish/jimmer/sql/api/BookService.java
package org.babyfish.jimmer.sql.api;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.babyfish.jimmer.client.Api;
import org.babyfish.jimmer.client.FetchBy;
import org.babyfish.jimmer.client.ThrowsAll;
import org.babyfish.jimmer.sql.fetcher.Fetcher;
import org.babyfish.jimmer.sql.model.AuthorFetcher;
import org.babyfish.jimmer.sql.model.Book;
import org.babyfish.jimmer.sql.model.BookFetcher;
import org.babyfish.jimmer.sql.model.BookStoreFetcher;
import org.babyfish.jimmer.sql.model.dto.BookInput;
import org.babyfish.jimmer.sql.model.dto.BookSpecification2;
import org.babyfish.jimmer.sql.model.dto.CompositeBookInput;
import org.jetbrains.annotations.Nullable;

import java.math.BigDecimal;
import java.util.List;

/**
 * An interface to test generation of `META-INF/jimmer/client`
 */
@Api(groups = "abc")
public interface BookService {

    void initialize(ObjectMapper mapper);

    @Api
    void processTree(Tree<Book> tree);

    /**
     * Find books with simple format
     * @param name Optional value to filter `name`
     * @param edition Optional value to filter `edition`
     * @param minPrice Optional value to filter `price`
     * @param maxPrice Optional value to filter `price`
     * @param storeName Optional value to filter `store.name`
     * @param authorName Optional value to filter `Author.firstName` or `Author.lastName`
     * @return A list of books objects, with short associations
     */
    @ThrowsAll(SystemErrorCode.class)
    @Api
    List<? extends @FetchBy("SIMPLE_FETCHER") Book> findSimpleBooks(
            @Nullable String name,
            @Nullable Integer edition,
            @Nullable BigDecimal minPrice,
            @Nullable BigDecimal maxPrice,
            @Nullable String storeName,
            @Nullable String authorName
    );

    /**
     * Find books with simple format
     * @param name Optional value to filter `name`
     * @param edition Optional value to filter `edition`
     * @param minPrice Optional value to filter `price`
     * @param maxPrice Optional value to filter `price`
     * @param storeName Optional value to filter `store.name`
     * @param authorName Optional value to filter `Author.firstName` or `Author.lastName`
     * @return A list of books objects, with long associations
     */
    @ThrowsSystemError({SystemErrorCode.A, SystemErrorCode.B})
    @Api
    List<@FetchBy("COMPLEX_FETCHER") Book> findComplexBooks(
            @Nullable String name,
            @Nullable Integer edition,
            @Nullable BigDecimal minPrice,
            @Nullable BigDecimal maxPrice,
            @Nullable String storeName,
            @Nullable String authorName
    );

    @Api
    List<? extends @FetchBy("COMPLEX_FETCHER") Book> findBySuperQBE(
            @Nullable BookSpecification2 specification
    );

    @Api
    @FetchBy("COMPLEX_FETCHER") Book findByNameAndEdition(
            String name,
            int edition
    );

    @Api
    void saveBook(BookInput input);

    @Api
    void saveBook(CompositeBookInput input);

    Fetcher<Book> SIMPLE_FETCHER =
            BookFetcher.$
                    .allScalarFields()
                    .store()
                    .authors();

    BookFetcher COMPLEX_FETCHER =
            BookFetcher.$
                    .allScalarFields()
                    .store(
                            BookStoreFetcher.$.allScalarFields()
                    )
                    .authors(
                            AuthorFetcher.$.allScalarFields()
                    );
}

File: project/jimmer-sql-kotlin/src/test/kotlin/org/babyfish/jimmer/sql/kt/api/BookService.kt
package org.babyfish.jimmer.sql.kt.api

import com.fasterxml.jackson.databind.ObjectMapper
import org.babyfish.jimmer.client.Api
import org.babyfish.jimmer.client.FetchBy
import org.babyfish.jimmer.client.ThrowsAll
import org.babyfish.jimmer.sql.fetcher.Fetcher
import org.babyfish.jimmer.sql.kt.fetcher.newFetcher
import org.babyfish.jimmer.sql.kt.model.classic.book.Book
import org.babyfish.jimmer.sql.kt.model.classic.book.by
import org.babyfish.jimmer.sql.kt.model.classic.book.dto.BookInput
import org.babyfish.jimmer.sql.kt.model.classic.book.dto.BookSpecification2
import org.babyfish.jimmer.sql.kt.model.classic.book.dto.CompositeBookInput
import java.math.BigDecimal

/**
 * An interface to test generation of `META-INF/jimmer/client`
 */
@Api(groups = ["abc"])
interface BookService {

    fun initialize(objectMapper: ObjectMapper)

    @Api
    fun handleTree(tree: Tree<Book>)

    /**
     * Find books with simple format
     * @param name Optional value to filter `name`
     * @param edition Optional value to filter `edition`
     * @param minPrice Optional value to filter `price`
     * @param maxPrice Optional value to filter `price`
     * @param storeName Optional value to filter `store.name`
     * @param authorName Optional value to filter `Author.firstName` or `Author.lastName`
     * @return A list of books objects, with long associations
     */
    @ThrowsAll(SystemErrorCode::class)
    @Api
    fun findSimpleBooks(
        name: String?,
        edition: Int?,
        minPrice: BigDecimal?,
        maxPrice: BigDecimal?,
        storeName: String?,
        authorName: String?
    ): List<@FetchBy("SIMPLE_FETCHER") Book>

    /**
     * Find books with simple format
     * @param name Optional value to filter `name`
     * @param edition Optional value to filter `edition`
     * @param minPrice Optional value to filter `price`
     * @param maxPrice Optional value to filter `price`
     * @param storeName Optional value to filter `store.name`
     * @param authorName Optional value to filter `Author.firstName` or `Author.lastName`
     * @return A list of books objects, with long associations
     */
    @ThrowsSystemError(SystemErrorCode.A, SystemErrorCode.B)
    @Api
    fun findComplexBooks(
        name: String?,
        edition: Int,
        minPrice: BigDecimal?,
        maxPrice: BigDecimal?,
        storeName: String?,
        authorName: String?
    ): List<@FetchBy("COMPLEX_FETCHER") Book>

    @Api
    fun findBySuperQBE(
        specification: BookSpecification2
    ): MutableList<out @FetchBy("COMPLEX_FETCHER") Book>

    @Api
    fun findByNameAndEdition(
        name: String,
        edition: Int
    ): @FetchBy("COMPLEX_FETCHER") Book?

    @Api
    fun saveBook(input: BookInput)

    @Api
    fun saveBook(input: CompositeBookInput)

    companion object {

        val SIMPLE_FETCHER: Fetcher<Book> =
            newFetcher(Book::class).by {
                allScalarFields()
                store()
                authors()
            }

        val COMPLEX_FETCHER =
            newFetcher(Book::class).by {
                allScalarFields()
                store()
                authors()
            }
    }
}